README.md
# KinoWardley

This is a livebook smart cell that contains a basic renderer for a Wardley Map

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `kino_wardley` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:kino_wardley, "~> 0.1.0"}
  ]
end
```

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at <https://hexdocs.pm/kino_wardley>.

This is a sample wardley map that can be generated using the following code:
```
KinoWardley.Output.new("""
id myid8
height 400
width 800
title Tea Shop
anchor Business [0.95, 0.63]
anchor Public [0.95, 0.78]
component Cup of Tea [0.79, 0.61] label [19, -4]
component Cup [0.73, 0.78]
component Tea [0.63, 0.81]
component Hot Water [0.52, 0.80]
component Water [0.38, 0.82]
component Kettle [0.43, 0.35] label [-57, 4]
evolve Kettle [0.43, 0.62] label [16, 7]
component Power [0.1, 0.7] label [-27, 20]
evolve Power [0.1, 0.89] label [-12, 21]
Business->Cup of Tea
Public->Cup of Tea
Cup of Tea->Cup
Cup of Tea->Tea
Cup of Tea->Hot Water
Hot Water->Water
Hot Water->Kettle 
Kettle->Power
evolution Experiment->Prototype->Production->Product
""")
```

![sample wardley map](https://raw.githubusercontent.com/chriseyre2000/kino_wardley/main/cup-of-tea.png "Sample Wardley Map")



Since large parts of this are in javascript it has a javascript test suite.

The test folder has a package.json.

`npm run test`

Here is a sample livebook that can be directly opened:

[![Run in Livebook](https://livebook.dev/badge/v1/gray.svg)](https://livebook.dev/run?url=https%3A%2F%2Fgithub.com%2Fchriseyre2000%2Flivebooks%2Fblob%2Fmain%2Fphoenix-wardley-map.livemd)



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.8.0]

### Added buy, build, outsource


## [0.7.0]

### Added inertia to Components
### Added some new Wardley Map components based upon a tweet (https://twitter.com/swardley/status/1565744194475499521)
### Added Market
### Fixed documentation of labels.

## [0.6.0]

### Adding style wardley
### diverging from onlinewardleymaps, adding a group

This is a typical group:
```
group first [0.605, 0.85, 20, 50] yellow black
```
The idea is that you can add ovals with various colours to group things together.
Overlapping ovals (without a distinct border) should be able to create any required
grouping.
### Added some documentation.

## [0.5.0]

### Now supports title
### Also supports note
### Now has annotations
### and the annotation details

## [0.4.1]

### Now all existing components are compatible with onlinewardleymaps syntax

## [0.4.0]

### Added unit tests
### Added a changelog
### Added the ability to define offsets as per onlinewardleymap.com

This allows things like:
`component Kettle [0.43, 0.35] label [-57, 4]`

I have not yet solved the evolve parsing.

### Added alternative axis.
```
KinoWardley.Output.new("""
id myid8
height 400
width 800
title Tea Shop
anchor Business [0.95, 0.63]
anchor Public [0.95, 0.78]
component Cup of Tea [0.79, 0.61] label [19, -4]
component Cup [0.73, 0.78]
component Tea [0.63, 0.81]
component Hot Water [0.52, 0.80]
component Water [0.38, 0.82]
component Kettle [0.43, 0.35] label [-57, 4]
evolve Kettle [0.43, 0.62] label [16, 7]
component Power [0.1, 0.7] label [-27, 20]
evolve Power [0.1, 0.89] label [-12, 21]
Business->Cup of Tea
Public->Cup of Tea
Cup of Tea->Cup
Cup of Tea->Tea
Cup of Tea->Hot Water
Hot Water->Water
Hot Water->Kettle 
Kettle->Power
evolution Experiment->Prototype->Production->Product
""")
```

## [0.3.0] - 2022-09-03
### Improved parsing via regex.
### Internally split the js file out of the ex file.


test/group_parse.test.js
import { parseSpec } from "../lib/assets/js/main"

test('Sample group spec parses to expected', () => {

  const input = `
  height 400
  width 800
  group My Group [0.5, 0.6, 30, 40] red black
  `
  const expected = {
    "My Group": {
      "cx": 0.6,
      "cy": 0.5,
      "name": "My Group",
      "type": "group",
      "rx": 40,
      "ry": 30,
      "fill": "red",
      "border": "black"
    },
    "__height": "400",
    "__width": "800",
  }

  expect(parseSpec(input.split("\n"))).toStrictEqual(expected);
});



test/evolve_parse.test.js
import { parseSpec } from "../lib/assets/js/main"

test('Sample evolve spec parses to expected', () => {

  const input = `
  id myid7
  height 400
  width 800
  component Kettle [0.43, 0.35]
  evolve Kettle [0.43, 0.62]
  component Power [0.10, 0.71]
  evolve Power [0.10, 0.89]
  Kettle->Power`

  const expected = {
    "Kettle": {
      "dx": 0,
      "dy": 0,
      "name": "Kettle",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "x": 0.35,
      "y": 0.43,
    },
    "Kettle__evolve": {
      "dx": 0,
      "dy": 0,
      "name": "Kettle",
      "type": "evolve",
      "x": 0.62,
      "y": 0.43,
    },
    "Power": {
      "dx": 0,
      "dy": 0,
      "name": "Power",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "x": 0.71,
      "y": 0.1,
    },
    "Power__evolve": {
      "dx": 0,
      "dy": 0,
      "name": "Power",
      "type": "evolve",
      "x": 0.89,
      "y": 0.1,
    },
    "__height": "400",
    "__id": "myid7",
    "__width": "800",
  }

  expect(parseSpec(input.split("\n"))).toStrictEqual(expected);
});

test('Online wardley evolve spec parses to expected', () => {

  const input = `
component Kettle [0.43, 0.35] label [-57, 4]
evolve Kettle 0.62 label [16, 7]
component Power [0.1, 0.7] label [-27, 20]
evolve Power 0.89 label [-12, 21]
`

  const expected = {
    "Kettle": {
      "dx": -57,
      "dy": 4,
      "name": "Kettle",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "x": 0.35,
      "y": 0.43,
    },
    "Kettle__evolve": {
      "dx": 16,
      "dy": 7,
      "name": "Kettle",
      "type": "evolve",
      "x": 0.62,
      "y": 0.43,
    },
    "Power": {
      "dx": -27,
      "dy": 20,
      "name": "Power",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "x": 0.7,
      "y": 0.1,
    },
    "Power__evolve": {
      "dx": -12,
      "dy": 21,
      "name": "Power",
      "type": "evolve",
      "x": 0.89,
      "y": 0.1,
    },
  }


  expect(parseSpec(input.split("\n"))).toStrictEqual(expected);

});

test('Simplified Online wardley evolve spec parses to expected', () => {

  const input = `
  component Kettle [0.43, 0.35]
  evolve Kettle 0.62
  component Power [0.1, 0.7]
  evolve Power 0.89
  `

  const expected = {
    "Kettle": {
      "dx": 0,
      "dy": 0,
      "name": "Kettle",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "x": 0.35,
      "y": 0.43,
    },
    "Kettle__evolve": {
      "dx": 0,
      "dy": 0,
      "name": "Kettle",
      "type": "evolve",
      "x": 0.62,
      "y": 0.43,
    },
    "Power": {
      "dx": 0,
      "dy": 0,
      "name": "Power",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "x": 0.7,
      "y": 0.1,
    },
    "Power__evolve": {
      "dx": 0,
      "dy": 0,
      "name": "Power",
      "type": "evolve",
      "x": 0.89,
      "y": 0.1,
    },
  }

  expect(parseSpec(input.split("\n"))).toStrictEqual(expected);

});

test/kino_wardley.test.js
import { parseSpec, extractNameLocationAndOffset } from "../lib/assets/js/main"

test('Empty spec is empty', () => {
  expect(parseSpec([])).toStrictEqual({});
});

test('Handle market', () => {
  const input = `
  market PCW [0.8, 0.5] inertia label [-27, -12]
  `

  const expected =    
  {
   "PCW":  {
     "dx": -27,
     "dy": -12,
     "inertia": true,
     "name": "PCW",
     "type": "market",
     "x": 0.5,
     "y": 0.8,
   },
 }

expect(parseSpec(input.split("\n"))).toStrictEqual(expected);

});

test('Handle inertia', () => {
  const input = `
  component Customer [0.95, 0.5] inertia label [-27, -12]
  `

  const expected =    
      {
       "Customer":  {
         "dx": -27,
         "dy": -12,
         "inertia": true,
         "size": "normal",
         "xor": false,
         "name": "Customer",
         "type": "component",
         "x": 0.5,
         "y": 0.95,
       },
     }

  expect(parseSpec(input.split("\n"))).toStrictEqual(expected);
})

test('Handle xor', () => {
  const input = `
  component Customer [0.95, 0.5] xor label [-27, -12]
  `

  const expected =    
      {
       "Customer":  {
         "dx": -27,
         "dy": -12,
         "inertia": false,
         "size": "normal",
         "xor": true,
         "name": "Customer",
         "type": "component",
         "x": 0.5,
         "y": 0.95,
       },
     }

  expect(parseSpec(input.split("\n"))).toStrictEqual(expected);
})

test('Sample spec parses to expected', () => {

  const input = `
  id myid7
  height 400
  width 800
  anchor Business [0.95, 0.63]
  anchor Public [0.95, 0.78]
  component Cup of Tea [0.79, 0.61]
  component Cup [0.73, 0.78]
  component Tea [0.63, 0.81]
  component Hot Water [0.52, 0.80]
  component Water [0.38, 0.82]
  component Kettle [0.43, 0.35]
  evolve Kettle [0.43, 0.62]
  component Power [0.10, 0.71]
  evolve Power [0.10, 0.89]
  Business->Cup of Tea
  Public->Cup of Tea
  Cup of Tea->Cup
  Cup of Tea->Tea
  Cup of Tea->Hot Water
  Hot Water->Water
  Hot Water->Kettle
  Kettle->Power`

  const expected = {
    "Business": {
      "name": "Business",
      "type": "anchor",
      "dx": 0,
      "dy": 0,
      "x": 0.63,
      "y": 0.95,
    },
    "Cup": {
      "name": "Cup",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.78,
      "y": 0.73,
    },
    "Cup of Tea": {
      "name": "Cup of Tea",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.61,
      "y": 0.79,
    },
    "Hot Water": {
      "name": "Hot Water",
      "type": "component",
      "size": "normal",
      "inertia": false,
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.8,
      "y": 0.52,
    },
    "Kettle": {
      "name": "Kettle",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.35,
      "y": 0.43,
    },
    "Kettle__evolve": {
      "name": "Kettle",
      "type": "evolve",
      "dx": 0,
      "dy": 0,
      "x": 0.62,
      "y": 0.43,
    },
    "Power": {
      "name": "Power",
      "type": "component",
      "size": "normal",
      "inertia": false,
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.71,
      "y": 0.1,
    },
    "Power__evolve": {
      "name": "Power",
      "type": "evolve",
      "dx": 0,
      "dy": 0,
      "x": 0.89,
      "y": 0.1,
    },
    "Public": {
      "name": "Public",
      "type": "anchor",
      "dx": 0,
      "dy": 0,
      "x": 0.78,
      "y": 0.95,
    },
    "Tea": {
      "name": "Tea",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.81,
      "y": 0.63,
    },
    "Water": {
      "name": "Water",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.82,
      "y": 0.38,
    },
    "__height": "400",
    "__id": "myid7",
    "__width": "800",
  }

  expect(parseSpec(input.split("\n"))).toStrictEqual(expected);
});

test('Online wardley map parses to expected', () => {

  const input = `
  id myid7
  height 400
  width 800
  title Tea Shop
  anchor Business [0.95, 0.63]
  anchor Public [0.95, 0.78]
  component Cup of Tea [0.79, 0.61] label [19, -4]
  component Cup [0.73, 0.78]
  component Tea [0.63, 0.81]
  component Hot Water [0.52, 0.80]
  component Water [0.38, 0.82]
  component Kettle [0.43, 0.35] label [-57, 4]
  evolve Kettle 0.62 label [16, 7]
  component Power [0.1, 0.7] label [-27, 20]
  evolve Power 0.89 label [-12, 21]
  Business->Cup of Tea
  Public->Cup of Tea
  Cup of Tea->Cup
  Cup of Tea->Tea
  Cup of Tea->Hot Water
  Hot Water->Water
  Hot Water->Kettle 
  Kettle->Power
  
  annotation 1 [[0.43,0.49],[0.08,0.79]] Standardising power allows Kettles to evolve faster
  annotation 2 [0.48, 0.85] Hot water is obvious and well known
  annotations [0.60, 0.02]
  
  note +a generic note appeared [0.23, 0.33]
  
  style wardley`

  const expected = {
    "__title": "Tea Shop",
    "Business": {
      "name": "Business",
      "type": "anchor",
      "dx": 0,
      "dy": 0,
      "x": 0.63,
      "y": 0.95,
    },
    "Cup": {
      "name": "Cup",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.78,
      "y": 0.73,
    },
    "Cup of Tea": {
      "name": "Cup of Tea",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": 19,
      "dy": -4,
      "x": 0.61,
      "y": 0.79,
    },
    "Hot Water": {
      "name": "Hot Water",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.8,
      "y": 0.52,
    },
    "Kettle": {
      "name": "Kettle",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": -57,
      "dy": 4,
      "x": 0.35,
      "y": 0.43,
    },
    "Kettle__evolve": {
      "name": "Kettle",
      "type": "evolve",
      "dx": 16,
      "dy": 7,
      "x": 0.62,
      "y": 0.43,
    },
    "Power": {
      "name": "Power",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": -27,
      "dy": 20,
      "x": 0.7,
      "y": 0.1,
    },
    "Power__evolve": {
      "name": "Power",
      "type": "evolve",
      "dx": -12,
      "dy": 21,
      "x": 0.89,
      "y": 0.1,
    },
    "Public": {
      "name": "Public",
      "type": "anchor",
      "dx": 0,
      "dy": 0,
      "x": 0.78,
      "y": 0.95,
    },
    "Tea": {
      "name": "Tea",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.81,
      "y": 0.63,
    },
    "Water": {
      "name": "Water",
      "type": "component",
      "inertia": false,
      "size": "normal",
      "xor": false,
      "dx": 0,
      "dy": 0,
      "x": 0.82,
      "y": 0.38,
    },
    "1": {
      "description": "Standardising power allows Kettles to evolve faster",
      "type": "annotation",
      "label": "1",
      "points": [
        {
          "x": 0.49,
          "y": 0.43,
        },
        {
          "x": 0.79,
          "y": 0.08,
        },
      ],
    },
    "2": {
      "description": "Hot water is obvious and well known",
      "label": "2",
      "type": "annotation",
      "points": [
        {
          "x": 0.85,
          "y": 0.48,
        },
      ],
    },

    "+a generic note appeared": {
      "dx": 0,
      "dy": 0,
      "name": "+a generic note appeared",
      "type": "note",
      "x": 0.33,
      "y": 0.23,
    },
    "__annotations": {
      "type": "annotations",
      "x": 0.02,
      "y": 0.6,
    },
    "__style": "wardley",
    "__height": "400",
    "__id": "myid7",
    "__width": "800",
  }
  expect(parseSpec(input.split("\n"))).toStrictEqual(expected);
});

test("regex extract full", () => {
  const output = extractNameLocationAndOffset("component Cup of Tea [0.79, 0.61] label [10, 11]")
  const expected = {
    "first": "0.79",
    "fourth": "11",
    "name": "Cup of Tea ",
    "second": "0.61",
    "third": "10",
  }
  expect(output).toEqual(expected);
})

test("regex extract short", () => {
  const output = extractNameLocationAndOffset("component Cup of Tea [0.79, 0.61]")
  const expected = {
    "first": "0.79",
    "fourth": "",
    "name": "Cup of Tea ",
    "second": "0.61",
    "third": "",
  }
  expect(output).toEqual(expected);
})

test("regex working test", () => {
  const output = /\w+\s(?<name>[^\[]*)\[(?<first>[^\,]*),\s*(?<second>[^\]]*)]\s*[label\s\[]*(?<third>[^,]*)[,\s]*(?<fourth>[^\]]*)/.exec("component Cup of Tea [0.79, 0.61] label [10, 11]").groups
  const expected = {
    "first": "0.79",
    "fourth": "11",
    "name": "Cup of Tea ",
    "second": "0.61",
    "third": "10",
  }
  expect(output).toEqual(expected);
})

test("parseEvolutionAxis", () => {
  const input = "evolution First->Second->Third->Fourth"

  const expected = {
    __evolution: [
      "First",
      "Second",
      "Third",
      "Fourth"
    ]
  }

  expect(parseSpec(input.split("\n"))).toStrictEqual(expected);
})



mix.exs
defmodule KinoWardley.MixProject do
  use Mix.Project

  def project do
    [
      app: :kino_wardley,
      version: "0.8.0",
      elixir: "~> 1.13",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      package: package(),
      docs: [
        extras: ["README.md", "Syntax.md", "CHANGELOG.md"]
      ]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      mod: {Kino.Wardley.Application, []}
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:kino, "~> 0.6.2"},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false}
    ]
  end

  defp package do
    [
      name: "kino_wardley",
      description: "A utility for rendering Wardley Maps in a Livebook.",
      files: ~w(lib lib/assets/js .formatter.exs cup-of-tea.png mix.exs README* LICENSE* CHANGELOG.md),
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/chriseyre2000/kino_wardley"}
    ]
  end
end


lib/kino_wardley/output.ex
defmodule KinoWardley.Output do
  use Kino.JS, assets_path: "lib/assets/js"

  @moduledoc """
  This is the output used to render Wardley Maps.
  """

  @doc """
  Create a wardley map.

  Here is a sample spec:
  ```
    id myid7
    height 400
    width 800
    anchor Business [0.95, 0.63]
    anchor Public [0.95, 0.78]
    component Cup of Tea [0.79, 0.61]
    component Cup [0.73, 0.78]
    component Tea [0.63, 0.81]
    component Hot Water [0.52, 0.80]
    component Water [0.38, 0.82]
    component Kettle [0.43, 0.35]
    evolve Kettle [0.43, 0.62]
    component Power [0.10, 0.71]
    evolve Power [0.10, 0.89]
    Business->Cup of Tea
    Public->Cup of Tea
    Cup of Tea->Cup
    Cup of Tea->Tea
    Cup of Tea->Hot Water
    Hot Water->Water
    Hot Water->Kettle
    Kettle->Power
  ```
  """
  def new(spec) do
    Kino.JS.new(__MODULE__, spec)
  end
end


lib/kino_wardley.ex
defmodule Kino.Wardley do
  @moduledoc """
  This is a smart cell that can be used to add a Wardley Map to a page.

  Wardley Maps are a chain of needs (a value chain) plotted against
  an evolution axis.

  See https://en.wikipedia.org/wiki/Wardley_map for a description of what a wardley map is.

  They are typically used to discuss strategy.

  """

  use Kino.JS, assets_path: "lib/assets/js"
  use Kino.SmartCell, name: "Wardley Map"
  use Kino.JS.Live

  @doc """
  This is used to create the smart cell.
  """
  def new(spec) do
    Kino.JS.new(__MODULE__, spec, export_info_string: "wardley")
  end

  @setup """
  {
    "width": 400,
    "height": 800
  }
  """

  @impl true
  def to_source(_) do
    '''
    KinoWardley.Output.new("""
    id myid7
    height 400
    width 800
    anchor Business [0.95, 0.63]
    anchor Public [0.95, 0.78]
    component Cup of Tea [0.79, 0.61]
    component Cup [0.73, 0.78]
    component Tea [0.63, 0.81]
    component Hot Water [0.52, 0.80]
    component Water [0.38, 0.82]
    component Kettle [0.43, 0.35]
    evolve Kettle [0.43, 0.62]
    component Power [0.10, 0.71]
    evolve Power [0.10, 0.89]
    Business->Cup of Tea
    Public->Cup of Tea
    Cup of Tea->Cup
    Cup of Tea->Tea
    Cup of Tea->Hot Water
    Hot Water->Water
    Hot Water->Kettle
    Kettle->Power
    """)
    '''
  end

  @impl true
  def to_attrs(_), do: %{}

  @setup """
  id myid7
  height 400
  width 800
  anchor Business [0.95, 0.63]
  anchor Public [0.95, 0.78]
  component Cup of Tea [0.79, 0.61]
  component Cup [0.73, 0.78]
  component Tea [0.63, 0.81]
  component Hot Water [0.52, 0.80]
  component Water [0.38, 0.82]
  component Kettle [0.43, 0.35]
  evolve Kettle [0.43, 0.62]
  component Power [0.10, 0.71]
  evolve Power [0.10, 0.89]
  Business->Cup of Tea
  Public->Cup of Tea
  Cup of Tea->Cup
  Cup of Tea->Tea
  Cup of Tea->Hot Water
  Hot Water->Water
  Hot Water->Kettle
  Kettle->Power
  """

  @impl true
  def handle_connect(ctx) do
    {:ok, @setup, ctx}
  end
end
